app/core/interfaces/relational/m2m/interface.js

```diff
@@ -85,7 +85,7 @@ define([
           if (me.columnSchema.options.get('no_duplicates') === true) {
             var duplicated = false;
             me.relatedCollection.each(function (model) {
-              if (model.get('data').id === id) {
+              if (+model.get('data').id === +id) {
```

Comment:
I like this in some level, but hopefully we let the API returns the actual casted value, _if this is not working s expected_. In fact don't letting the user make relation between two different types.

Reply:
I'll double check the types, but based on the config in #1999, it looks like a bug.

Reply:
Okay, the reason why this happens is due to this line:
`_.each(view.table.selection(), function (id) {`
As it's just grabbing the table dom checked rows, grabbing the value from each row.  Being a dom element, we lose the type, so we just get a string.  
I guess the .selection method could perform the cast, but is there any situation where the value may be a non-number?
```
/app/core/table/table.view.js:

    selection: function() {
      var selection = [];

      this.$('td .js-select-row:checked').each(function () {
        selection.push(this.value);
      });

      return selection;
    },
```

Reply:
Yeah, the ID (primary key) can be any datatype. F.e. in a `languages` table, the ID can be the language code (`en_us`) instead of a number

Reply:
hmm, shall I cast to string for comparison? either that or drop the === to ==, but that feels sloppy.

Reply:
ping @rijkvanzanten @WellingGuzman any outcome to a solution to this? is m2m code used in the foreign language logic?  otherise I probably can swap it to forcing a comparison of strings?
`if (model.get('data').id.toString() === id.toString()) {`

Reply:
@jedimonkey Can you give us a example where the one of the was string? because this is probably a bug in the API then.

I see your concern, as this may solve the problem now, I'd like to fix it (if possible) from the root.

Reply:
From memory when I was debugging, I believe the issue is not in the api, but the fact it's coming from the dom id attribute.  It's always going to come back as a string, due to id's being strings. I considered casting it in that function (https://github.com/directus/directus/blob/6c0f8602cb53af4b6969a055b77f18113523a853/app/core/table/table.view.js#L67), but then that may break other checks that rely on it always returning strings.  Ultimately it needs to be cast somewhere.

Reply:
```js
if (+model.get('data').id === +id) {
```

Your logic sounds right! The ID comes from the DOM, making it a string. Casting it to a number should work for _numeric based IDs_. However...... string based IDs are a thing too!

I think it's the safest option to cast both to a String instead of a number, seeing that would prevent `NaN` for string based values 👍